From: Martin Pitt <martin.pitt@ubuntu.com>
Date: Fri, 3 Apr 2026 15:15:13 +0800
Subject: Deprecate pam_unix' explicit "usergroups" option and instead read it
 from /etc/login.def's "USERGROUP_ENAB" option if umask is only defined
 there. This restores compatibility with the pre-PAM behaviour of login. See
 https://blueprints.launchpad.net/ubuntu/+spec/umask-to-0002.

Bug-Debian: http://bugs.debian.org/583958
Last-Update: 2019-02-12

=== modified file 'modules/pam_umask/pam_umask.c'
---
 modules/pam_umask/pam_umask.8.xml |  8 +++++++-
 modules/pam_umask/pam_umask.c     | 16 ++++++++++++++++
 2 files changed, 23 insertions(+), 1 deletion(-)

diff --git a/modules/pam_umask/pam_umask.8.xml b/modules/pam_umask/pam_umask.8.xml
index 54a44ef..97c4b74 100644
--- a/modules/pam_umask/pam_umask.8.xml
+++ b/modules/pam_umask/pam_umask.8.xml
@@ -58,7 +58,8 @@
         </listitem>
         <listitem>
           <para>
-            UMASK entry from /etc/login.defs
+            UMASK entry from /etc/login.defs (influenced by USERGROUPS_ENAB in
+	    /etc/login.defs)
           </para>
         </listitem>
         <listitem>
@@ -117,6 +118,11 @@
               If the user is not root and the username is the same as
               primary group name, the umask group bits are set to be the
               same as owner bits (examples: 022 -&gt; 002, 077 -&gt; 007).
+	      Note that using this option explicitly is discouraged. pam_umask
+	      enables this functionality by default if /etc/login.defs enables
+	      USERGROUPS_ENAB, and the umask is not set explicitly in other
+	      places than /etc/login.defs (this is compatible with login's
+	      behaviour without PAM).
             </para>
           </listitem>
         </varlistentry>
diff --git a/modules/pam_umask/pam_umask.c b/modules/pam_umask/pam_umask.c
index 8ba5de9..9eebdd1 100644
--- a/modules/pam_umask/pam_umask.c
+++ b/modules/pam_umask/pam_umask.c
@@ -95,6 +95,8 @@ static int
 get_options (pam_handle_t *pamh, options_t *options,
 	     int argc, const char **argv)
 {
+  char *result;
+
   memset (options, 0, sizeof (options_t));
 
   options->usergroups = DEFAULT_USERGROUPS_SETTING;
@@ -105,6 +107,20 @@ get_options (pam_handle_t *pamh, options_t *options,
 
   if (options->umask == NULL) {
     options->login_umask = pam_modutil_search_key (pamh, LOGIN_DEFS, "UMASK");
+    /* login.defs' USERGROUPS_ENAB will modify the UMASK setting there by way
+     * of usergroups; but we don't want it to influence umask definitions
+     * from other places (like GECOS). This restores compatibility with
+     * shadow from the pre-PAM age.
+     */
+    if (options->login_umask != NULL)
+    {
+      result = pam_modutil_search_key (pamh, LOGIN_DEFS, "USERGROUPS_ENAB");
+      if (result != NULL)
+      {
+        options->usergroups = (strcasecmp (result, "yes") == 0);
+        free (result);
+      }
+    }
     if (options->login_umask == NULL)
       options->login_umask = pam_modutil_search_key (pamh, LOGIN_CONF, "UMASK");
     options->umask = options->login_umask;
